### PR TITLE
[AutoDiff] Fix `@differentiable` attribute SILGen and serialization.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Propagate `@differentiable` attributes in `SILFunctionBuilder::addFunctionAttributes`.
  - Previously, this logic was missing, causing SIL `[differentiable]` attributes
    to be missing during SILGen.
  - This unblocks using VJP definitions in the stdlib (with some extra work).
- Serialize `AutoDiffParameterIndices` for `DifferentiableAttr` and use it as the
 primary source for differentiation parameter info (instead of
  `ParsedAutoDiffParameter`).
- Rework differentiation to never load SIL functions.
  - Previously, differentiation relied on explicit loading to get SIL
    `[differentiable]` attributes. That was a hack.
  - Now that `@differentiable` attribute is propagated and serialized,
    such loading is no longer necessary.
- Tighten differentiation infrastructure to uphold the following invariants:
  - Differentiable functions' primal/adjoint are visible only in defining module.
  - Differentiable functions' JVP/VJPs are always visible to other modules.
    The differentiation pass guarantees that JVP/VJPs exist.

There are two `TensorFlowRuntime` test regressions in GPE mode.
- Both are because used device set includes RUNTIME device in
  `GraphFunctionDeviceInfo::finalizeUsedDevices()`.

Todos:
- Replace adjoint definitions in stdlib with VJP definitions.
- Eliminate primal/adjoint from `@differentiable` attribute.
  Use JVP/VJPs everywhere.

### CODE REVIEW

File: include/swift/AST/AutoDiff.h [hunk 1]
```diff
@@ -109,12 +111,13 @@ class AutoDiffParameterIndices : public llvm::FoldingSetNode {
...
   const llvm::SmallBitVector parameters;
 
-  AutoDiffParameterIndices(llvm::SmallBitVector parameters)
-      : parameters(parameters) {}
-
   static AutoDiffParameterIndices *get(llvm::SmallBitVector parameters,
                                        ASTContext &C);
 
+private:
+  AutoDiffParameterIndices(llvm::SmallBitVector parameters)
```
Comment (rxwei) [inline include/swift/AST/AutoDiff.h]:
```suggestion
  AutoDiffParameterIndices(const llvm::SmallBitVector &parameters)
```

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 2]
```diff
@@ -990,6 +990,7 @@ class ADContext {
   lookUpOrRegisterDifferentiationTask(SILFunction *original,
                                       const SILAutoDiffIndices &indices,
                                       DifferentiationInvoker invoker) {
+    /*
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: You can just delete this hunk.

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -2460,11 +2459,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   } else {
     // 'wrt:' is specified. Validate and collect the selected parameters.
     int lastIndex = -1;
-    for (size_t i = 0; i < uncheckedWrtParams.size(); i++) {
-      auto paramLoc = uncheckedWrtParams[i].getLoc();
-      switch (uncheckedWrtParams[i].getKind()) {
-      case AutoDiffParameter::Kind::Index: {
-        unsigned index = uncheckedWrtParams[i].getIndex();
+    for (size_t i = 0; i < parsedWrtParams.size(); i++) {
```
Comment (rxwei) [inline lib/Sema/TypeCheckAttr.cpp]: Some improvement:
```suggestion
    for (unsigned i : indices(parsedWrtParams)) {
```

File: lib/Serialization/Deserialization.cpp [hunk 4]
```diff
@@ -2586,21 +2586,17 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
...
-        SourceLoc loc;
-        for (auto paramValue : paramValues) {
-          auto parameter = paramValue & 0x01
-            ? AutoDiffParameter::getSelfParameter(loc)
-            : AutoDiffParameter::getIndexParameter(loc, paramValue >> 1);
-          parameters.push_back(parameter);
-        }
-        // TODO: Deserialize CheckedParameterIndices.
+        llvm::SmallBitVector parametersBitVector(parameters.size());
+        for (unsigned i = 0; i < parameters.size(); i++)
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
        for (unsigned i : indices(parameters))
```

File: test/AutoDiff/derived_differentiable_properties.swift [hunk 5]
```diff
@@ -7,7 +7,7 @@ public struct Foo : Differentiable {
 }
 
 // CHECK-AST-LABEL: public struct Foo : Differentiable {
-// CHECK-AST:   @sil_stored @differentiable()
+// CHECK-AST:   @sil_stored @differentiable(wrt: (self))
```
Comment (rxwei) [inline test/AutoDiff/derived_differentiable_properties.swift]: I love that things have a canonical printout now when you use checked parameter indices.

**Review by rxwei — APPROVED** [review]

File: lib/Sema/TypeCheckAttr.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/Sema/TypeCheckAttr.cpp]: Thanks! Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d.

File: lib/Serialization/Deserialization.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/Serialization/Deserialization.cpp]: Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d. Also done in `SerializeSIL.cpp` and `DeserializeSIL.cpp`.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Woo! Done.

File: include/swift/AST/AutoDiff.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/AutoDiff.h]: Done in https://github.com/apple/swift/pull/21837/commits/6fe2c6458258a66d31c72730d89aed36f297996d.

**Comment by dan-zheng** [comment]
@swift-ci Please clean test tensorflow macOS

**Comment by dan-zheng** [comment]
@swift-ci Please clean test tensorflow macOS

**Comment by dan-zheng** [comment]
@swift-ci Please clean test tensorflow macOS

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow Linux GPU

File: lib/Serialization/Serialization.cpp [hunk 6]
```diff
@@ -2369,24 +2369,16 @@ void Serializer::writeDeclAttribute(const DeclAttribute *DA) {
...
-      // Index parameters are left-shifted by 1.
-      case AutoDiffParameter::Kind::Index:
-        parameters.push_back(param.getIndex() << 1);
-        break;
-      }
-    }
+    auto paramIndices = attr->getParameterIndices();
+    assert(paramIndices && "Checked parameter indices must be resolved");
+    SmallVector<bool, 4> indices;
+    for (unsigned i = 0; i < paramIndices->parameters.size(); i++)
```
Comment (rxwei) [inline lib/Serialization/Serialization.cpp]:
```suggestion
    for (unsigned i : indices(paramIndices->parameters.size()))
```

**Review by rxwei — APPROVED** [review]

File: lib/Serialization/Serialization.cpp (hunk 6 above)
Comment (dan-zheng) [reply lib/Serialization/Serialization.cpp]: Thanks! Done in https://github.com/apple/swift/pull/21837/commits/5d8016b1850d755062fba2d9f19c12495569b5ac.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux

**Comment by dan-zheng** [comment]
This was joint work with @rxwei. 🤝

**Pull request merged**